pkg/libvirt/console.go

```diff
@@ -0,0 +1,216 @@
+package libvirt
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"sync"
+
+	"github.com/projecteru2/yavirt/pkg/log"
+	"github.com/projecteru2/yavirt/pkg/utils"
+	golibvirt "github.com/projecteru2/yavirt/third_party/libvirt"
+)
+
+type ConsoleFlags struct {
+	Force    bool
+	Safe     bool
+	Nonblock bool
+}
+
+func (cf *ConsoleFlags) genLibvirtFlags() (flags golibvirt.DomainConsoleFlags) {
+	if cf.Force {
+		flags |= golibvirt.DomainConsoleForce
+	}
+	if cf.Safe {
+		flags |= golibvirt.DomainConsoleSafe
+	}
+	return
+}
+
+func (cf *ConsoleFlags) genStreamFlags() (flags golibvirt.StreamFlags) {
+	if cf.Nonblock {
+		flags = golibvirt.StreamNonblock
+	}
+	return
+}
+
+type Console struct {
+	Stream *Stream
+	// pty to user
+	fromQ *utils.BytesQueue
+	// user to pty
+	toQ *utils.BytesQueue
+
+	quit chan struct{}
+	once sync.Once
+}
+
+func newConsole(s *Stream) *Console {
+	return &Console{
+		Stream: s,
+		fromQ:  utils.NewBytesQueue(),
+		toQ:    utils.NewBytesQueue(),
+		quit:   make(chan struct{}),
+	}
+}
+
+func (c *Console) needExit(ctx context.Context) bool {
+	if ctx == nil {
+		ctx = context.Background()
+	}
+	select {
+	case <-ctx.Done():
+		return true
+	case <-c.quit:
+		return true
+	default:
+		return false
+	}
+}
+
+func (c *Console) From(ctx context.Context, r io.Reader) error {
+	buf := make([]byte, 64*1024)
+	for {
+		if c.needExit(ctx) {
+			return nil
+		}
+		// Read a single byte
+		n, err := r.Read(buf)
+		if err != nil {
+			if err != io.EOF {
+				log.Errorf("[Console:From] read error: %s", err)
+			}
+			return err
+		}
+
+		if n == 0 {
+			continue
+		}
+
+		bs := buf[:n]
+		cloneBs := bytes.Clone(bs)
+		copy(buf, make([]byte, len(buf)))
```

Comment:
does it mean clear from the `buf`, if so, is clear the leftmost `n` enough?

Reply:
yes, it does. because each message is received by the stream one by one, if does not so, the buf may left dirty data for the next loop.